### PR TITLE
builtin/read: implement field splitting

### PR DESCRIPTION
**Pull request by delthas**
also print PS2 on line continuation

### CODE REVIEW

File: builtin/read.c [hunk 1]
```diff
@@ -47,6 +41,7 @@ int builtin_read(struct mrsh_state *state, int argc, char *argv[]) {
 		if (c == '\n') {
 			if (escaped) {
 				escaped = false;
+				fprintf(stderr, "%s", mrsh_env_get(state, "PS2", NULL));
```
Comment (emersion) [inline builtin/read.c]: This crashes if `PS2` is unset.

File: builtin/read.c [hunk 2]
```diff
@@ -55,8 +51,36 @@ int builtin_read(struct mrsh_state *state, int argc, char *argv[]) {
...
+	struct mrsh_word_string *str = mrsh_word_string_create(mrsh_buffer_steal(&buf), false);
+	
+	struct mrsh_array fields = {0};
+	split_fields(&fields, &str->word, mrsh_env_get(state, "IFS", NULL));
 
-	mrsh_env_set(state, argv[optind], buf.data, MRSH_VAR_ATTRIB_NONE);
-	mrsh_buffer_finish(&buf);
+	if (fields.len <= argc - optind) {
+		for (size_t i = 0; i < fields.len; ++i) {
+			mrsh_env_set(state, argv[optind + i], (char*)fields.data[i], MRSH_VAR_ATTRIB_NONE);
```
Comment (emersion) [inline builtin/read.c]: Style: missing space `(char *)`

File: builtin/read.c [hunk 3]
```diff
@@ -55,8 +51,36 @@ int builtin_read(struct mrsh_state *state, int argc, char *argv[]) {
 		mrsh_buffer_append_char(&buf, (char)c);
 	}
 	mrsh_buffer_append_char(&buf, '\0');
+	
+	struct mrsh_word_string *str = mrsh_word_string_create(mrsh_buffer_steal(&buf), false);
```
Comment (emersion) [inline builtin/read.c]: Nitpick: is it possible to rename this to `ws`? It's too easy to assume it's just a `char *`.

File: builtin/read.c [hunk 4]
```diff
@@ -55,8 +51,36 @@ int builtin_read(struct mrsh_state *state, int argc, char *argv[]) {
...
+			if (i != fields.len - 1) {
+				// TODO add the field delimiter rather than space (bash and dash always use spaces)
+				mrsh_buffer_append_char(&buf_last, ' ');
+			}
+		}
+		mrsh_env_set(state, argv[argc - 1], mrsh_buffer_steal(&buf_last), MRSH_VAR_ATTRIB_NONE);
+	}
+	
+	mrsh_word_destroy(&str->word);
+	mrsh_array_finish(&fields);
```
Comment (emersion) [inline builtin/read.c]: Each field is a `char *` that needs to be free'd. We need to iterate over `fields` to free each item.

File: builtin/read.c [hunk 5]
```diff
@@ -55,8 +51,36 @@ int builtin_read(struct mrsh_state *state, int argc, char *argv[]) {
...
+		struct mrsh_buffer buf_last = {0};
+		for (int i = argc - optind - 1; i < fields.len; ++i) {
+			char *field = (char*)fields.data[i];
+			mrsh_buffer_append(&buf_last, field, strlen(field));
+			if (i != fields.len - 1) {
+				// TODO add the field delimiter rather than space (bash and dash always use spaces)
+				mrsh_buffer_append_char(&buf_last, ' ');
+			}
+		}
+		mrsh_env_set(state, argv[argc - 1], mrsh_buffer_steal(&buf_last), MRSH_VAR_ATTRIB_NONE);
```
Comment (emersion) [inline builtin/read.c]: We need to add a terminating `\0` byte to the buffer before stealing it.

**Comment by emersion** [comment]
I just get a few signed comparison errors when compiling:

```
../builtin/read.c: In function ‘builtin_read’:
../builtin/read.c:61:17: error: comparison of integer expressions of different signedness: ‘size_t’ {aka ‘long unsigned int’} and ‘int’ [-Werror=sign-compare]
  if (fields.len <= argc - optind) {
                 ^~
../builtin/read.c:65:33: error: comparison of integer expressions of different signedness: ‘size_t’ {aka ‘long unsigned int’} and ‘int’ [-Werror=sign-compare]
   for (size_t i = fields.len; i < argc - optind; ++i) {
                                 ^
../builtin/read.c:73:37: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
   for (int i = argc - optind - 1; i < fields.len; ++i) {
                                     ^
../builtin/read.c:76:10: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
    if (i != fields.len - 1) {

```

**Comment by emersion** [comment]
Here's a small memory leak:

```
==22340== Memcheck, a memory error detector
==22340== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==22340== Using Valgrind-3.14.0.GIT and LibVEX; rerun with -h for copyright info
==22340== Command: build/mrsh
==22340== 
$ read a
a b c
$ echo $a
a b c
$ 
==22340== 
==22340== HEAP SUMMARY:
==22340==     in use at exit: 96 bytes in 2 blocks
==22340==   total heap usage: 311 allocs, 309 frees, 11,573 bytes allocated
==22340== 
==22340== 32 bytes in 1 blocks are definitely lost in loss record 1 of 2
==22340==    at 0x4838425: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==22340==    by 0x483C08C: realloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==22340==    by 0x48532C1: mrsh_buffer_reserve (buffer.c:15)
==22340==    by 0x4853321: mrsh_buffer_add (buffer.c:28)
==22340==    by 0x4853377: mrsh_buffer_append (buffer.c:38)
==22340==    by 0x4854FF8: builtin_read (read.c:75)
==22340==    by 0x4853E60: mrsh_run_builtin (builtin.c:65)
==22340==    by 0x485D6E6: task_builtin_poll (command.c:122)
==22340==    by 0x485DD29: task_command_poll (command.c:277)
==22340==    by 0x485E7BD: task_poll (task.c:30)
==22340==    by 0x485DF8C: task_list_poll (list.c:27)
==22340==    by 0x485E7BD: task_poll (task.c:30)
==22340== 
==22340== LEAK SUMMARY:
==22340==    definitely lost: 32 bytes in 1 blocks
==22340==    indirectly lost: 0 bytes in 0 blocks
==22340==      possibly lost: 0 bytes in 0 blocks
==22340==    still reachable: 64 bytes in 1 blocks
==22340==         suppressed: 0 bytes in 0 blocks
==22340== Reachable blocks (those to which a pointer was found) are not shown.
==22340== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==22340== 
==22340== For counts of detected and suppressed errors, rerun with: -v
==22340== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

**Comment by delthas** [comment]
is it good now? i cant easily check for memleaks with my current toolchain

**Comment by emersion** [comment]
Thanks!

**Pull request merged**